Lib/logging/__init__.py

```diff
@@ -458,29 +462,38 @@ class StrFormatStyle(PercentStyle):
     asctime_format = '{asctime}'
     asctime_search = '{asctime'
     # This is only for validating the format_spec
-    validation_pattern = re.compile(r"^((.)?[<>=]?[+ -]?[#0]?[\d]*,?[\.\d]*)?[bcdefgnosx%]?$", re.I)
+    validation_pattern = re.compile(r"^(.?[<>=^]?[+ -]?#?0?[\d]*,?(\.[\d]*)?)?[bcdefgnosx%]?$", re.I)
```

Comment:
I had a bit of a think about this whole area - I think both brace and dollar validation need a bit more work. I put my thoughts into [this Gist](https://gist.github.com/vsajip/d1b135010a07b7cd35673544feb61cd8) - see what you think and if you can adapt some of it. The test cases in there might be useful, too - and you could try adding to them using combinations I haven't tried.

I also think maybe we should add a `validate=` keyword parameter to `Formatter.__init__`, which defaults to `True` but which a user can set to `False` to disable validation, as a workaround for if we miss something.

It means that the `validation_pattern` idea I had was inadequate - for %-formatting a single pattern seems OK, for {-formatting we need the two-step approach you came up with, and I also realised that we hadn't catered for the no-fields case for $-formatting. But there's little point in having a uniform `validate()` that just relies on one pattern, as that won't fly. :disappointed: